README.md
Bibliostratus : STRatégie d'Alignement d'URIs pour la Transition bibliographique
==

[![alt Télécharger Bibliostratus](https://raw.githubusercontent.com/Transition-bibliographique/bibliostratus/master/img/bouton_telecharger_bibliostratus.png)](https://github.com/Transition-bibliographique/bibliostratus/tree/master/bin)

* Le répertoire *bibliostratus* contient le code développé en Python
* Le répertoire *bin* contient la dernière version compilée de l'exécutable, à télécharger et lancer directement pour utiliser le programme (le code source n'est pas forcément compilé à chaque modification)

Bibliostratus a pour objectif de proposer un ensemble de fonctions permettant de s'aligner avec les données bibliographiques de la [Bibliothèque nationale de France](http://www.bnf.fr), afin de pouvoir bénéficier de leurs travaux de FRBRisation de masse.

Actuellement quatre modules sont en cours de développement, non encore finalisés :

* **marc2tables** : permet de charger en entrée l'extraction d'un catalogue de bibliothèque en Unimarc (formats iso2709 ou MarcXML). Le programme produit autant de fichiers tabulés (tableaux) qu'il y a de combinaisons Type de document / type de notice. Ces tableaux sont au format attendu pour le 2e programme
* **noticesbib2arkBnF** : exploitation des métadonnées de chaque notice bibliographique pour identifier la notice correspondante dans le catalogue BnF. Pour l'instant, on charge les ressources par type de document/notice (monographies imprimées, CD/DVD, périodiques, etc.). 
* **noticesaut2arkBnF** : son équivalent pour s'aligner avec les autorités (Personnes physiques et organisations), peu testé pour le moment
* **ark2records** : en entrée un fichier listant des ARK BnF, afin de récupérer les notices bibliographiques (+option avec notices d'autorité) correspondantes. Le format en sortie est pour l'instant du XML. Il devrait y avoir de l'iso2709 à terme.

Le fichier **main** encapsule les 4 précédents pour les concevoir comme autant d'étapes dans la transition bibliographique des établissements

Objectifs du programme
--

Le présupposé de ce projet est que les bibliothèques doivent, à terme, bénéficier de catalogues contenant des notices conformes au modèle FRBR-LRM. 
Afin d'éviter qu'elles ne transforment leurs notices de manière indépendantes, ou qu'elles achètent les notices transformées à des fournisseurs (ou pire, qu'elles acquièrent des logiciels qui calculent le regroupement de notices en arbres FRBR à la volée), les agences nationales proposent de faciliter la redistribution de leurs notices.
Mais pour cela, il faut que chaque bibliothèque ait fait un travail préalable d'*alignement* de son catalogue avec la BnF ou l'Abes, pour que chaque notice d'un catalogue local soit "lié" à la notice du réservoir national (catalogue général BnF ou Sudoc)

Mode d'emploi et tutoriels
--
* [Voir le wiki](https://github.com/Transition-bibliographique/bibliostratus/wiki "Consulter les pages du wiki sur Github")
* [Consulter le forum utilisateurs](http://www.agorabib.fr/topic/3317-bibliostratus-mettre-en-correspondance-ses-notices-avec-celles-de-la-bnf/ "topic Agorabib")
* [Consulter la documentation technique](https://github.com/Transition-bibliographique/bibliostratus/tree/master/doc)
* [Installer Bibliostratus sur Linux](INSTALL.md)


source/release_notes.md
### version 1.21
juillet 2018
- Enrichissement des classes Bib_record et Aut_record
- Fichiers d'exemples consultables en local (et non en ligne)
- Module bleu : possibilité d'importer un fichier Marc21 (en modifiant le fichier main/files/preferences.json)
- favicône "Transition bibliographique" sur les fenêtres du module

###version 1.20
04/07/2018
- Tous modules : 
	- mise en forme PEP8 et flake8
	- initialisation d'un fichier de tests automatiques (à développer)
	- correction lignes lancement
- Module bleu (marc2tables):
	- option : modifier le fichier main\preferences.json pour avoir en entrée un fichier en marc21 plutôt qu'en unimarc
- Module blanc (alignement BIB):
	- Option BnF > Sudoc ou Sudoc > BnF
	- Création d'une classe Bib_record pour gérer les métadonnées en entrée
	Effets de bords possibles sur certaines parties du code (même si tous les impacts sont censés avoir été repris)
- Module rouge (ark2marc) :
	- extraction des notices à partir d'une liste de PPN (et pas seulement d'une liste d'ARK) ou d'une liste mélangée PPN + ARK
	


### version 1.19
08/06/2018
- Module bleu (marc > tables) :
	- option : exporter des métadonnées de notices bibliographiques en vue d'alignements avec des autorités (combinaison Titre-Auteur)
	- nouveau code AUD d'alignement pour les enregistrements sonores, distincts des enregistrements audiovisuels
	- Extraction biblio pour alignement d'autorités : ajout de la date de publication dans les critères d'alignement
- Module blanc (alignements bibliographiques):
	- Alignement sur l'ISSN : par expression exacte (option "adj" du SRU) et non tous mots (option "all")
	- Alignement Titre-Auteur-Date : contrôle sur la date pour vérifier qu'elle est bien présente en date d'édition (zone 100 ou 210$d).
	- fonction spécifique "par item". Facilite la prise en charge du même code par d'autres programmes (via API, interface web, etc.)
	- distinction entre enregistrements sonores (AUD) et vidéos (VID) : alignement sur l'un ou sur l'autre de ces deux types de documents
	- Contrôle sur la date : ajout de la date de copyright (302$a) en plus de la zone 100 et 210$d
- Module blanc (alignements autorités) :
	- Alignement à partir de métadonnées bibliographiques (combinaison Titre-Nom d'auteur) : ajout du critère "Date de publication"
	- Alignement à partir de métadonnées bibliographiques : Alignement sur nom de famille et sur prénom (et pas sur nom de famille seul)

### version 1.17
12/05/2018
Corrections diverses dans les formulaires : homogénéisation, etc.
- Module bleu (marc > tables) :
	- tables de ebooks générés comme les imprimés -> deviennent alignables
- Module blanc (Alignement AUT):
	- Si pas de date de début, alignement sur point d'accès + date de fin
- Module blanc (Alignements BIB et AUT):
	- Amélioration messages d'erreurs (colonnes manquantes en entrée)

### version 1.16
02/05/2018
- Module bleu (marc > tables):
	- Autorités : récupération des dates en 200$f si la zone à positions fixes 103 n'est pas renseignée
- Module blanc (alignement Autorités):
	- si un ISNI a été trouvé, le module vérifie si cet ISNI est dans le catalogue BnF. En effet la base ISNI connaît certaines formes d'auteurs absentes du catalogue BnF
	
### version 1.15
20/04/2018
- Tous modules :
	- Correction lien "Documentation en ligne"
- Module blanc (alignements BIB):
	- Quand la recherche ISBN n'a rien donné : relance d'une recherche ISBN partout dans la notice seulement si sa valeur nettoyée est d'au moins 10 caractères
	- Correction sur l'erreur ConnectionAbortedError
- Module blanc (alignements AUT):
	- nettoyage et formatage des ISNI avant requête dans le SRU
- Module rouge (marc > tableaux) :
	- Erreur d'encodage XML : message d'erreur + fichier en sortie
	- Prise en compte d'un fichier ISO2709 UTF-8 avec BOM : nettoyage du BOM (dans un fichier temporaire) avant d'extraire les notices

### version 1.14
17/04/2018
- Modules blancs (alignements) :
    - ajout d'un marqueur "&origin=bibliostratus" dans les requêtes sur le SRU BnF
	- bug : Exception ConnectionAbortedError mal levée

### version 1.13
16/04/2018
- Module blanc (alignement notices bibliographiques)
	- meilleur nettoyage de l'ISBN
- Module blanc (alignement notices d'autorité)
	- intégration code d'erreur 10053 (problème connexion à l'API)
- Module rouge (extraction de notices à partir d'une liste d'ARK):
	- correction pour extraction correcte d'ARK autorités
	
### version 1.12
14/04/2017
- Module blanc (alignements notices bibliographiques) : 
	- Recherche sur numéro commercial : si 0 résultat quand on cherche sur le critère bib.comref, le programme relançait la recherche partout dans la notice --> ce rebond est désactivé (le "numéro commercial" peut avoir des valeurs trop génériques, comme "208")
	- Correction bug sur les notices sans date pour les recherches Titre-Auteur-Date (la valeur vide de la date était remplacée par "*" qui faisait planter la requête SRU)

### version 1.11
- Module blanc (alignements notices bibliographiques) : 
	- Correction du format des fichiers en sortie quand l'option "Plusieurs fichiers est choisie"
	- Ajout d'un contrôle Titre sur l'alignement par numéro commercial
	- Correction de la fonction Nettoyage ISBN (les ponctuations, etc. n'étaient pas retirées en réalité jusque-là)


### version 1.10
**06/04/2018**

- Tous modules : ajout d'un bouton "Forum utilisateurs" > agoraBib
- Module bleu Marc > tables : 
	- récupération de toutes les autorités (certaines écrasées par d'autres)
	- rapidité : notices écrites directement dans les rapports en sortie au lieu de passer par une liste intermédiaire
	- Stats finales sur le terminal (volumétrie des notices ventilées)

### version 1.09
**04/04/2018**

- Module bleu Marc > tableaux : résolution d'autres problèmes d'encodage des fichiers en entrée

### version 1.08 
**01/04/2018**

- Module bleu Marc > tableaux : résolution des problèmes d'encodage :
désormais, les notices avec caractères non conformes UTF-8 ne font pas planter le programme

marc2tables.py
# -*- coding: utf-8 -*-
"""
Created on Fri Oct 13 18:30:30 2017

@author: Etienne Cavalié

Conversion de fichiers XML ou iso2709 en tableaux pour alignements

"""


from unidecode import unidecode
import urllib.error as error
import tkinter as tk
from tkinter import filedialog
from collections import defaultdict
import re
import webbrowser
import noticesbib2arkBnF as bib2ark
import pymarc as mc
import main as main

version = 0.01
programID = "marc2tables"
lastupdate = "10/11/2017"
last_version = [version, False]



# =============================================================================
# Creation des fichiers résultats
# =============================================================================
popup_filename = []

def create_file_doc_record(doc_record, id_traitement):
    filename= "-".join([id_traitement, doc_record_type[doc_record]]) + ".txt"
    file = open(filename, "w", encoding="utf-8")
    return file

# =============================================================================
# Fonctions de nettoyage
# =============================================================================
chiffers = ["0","1","2","3","4","5","6","7","8","9"]
letters = ["a","b","d","e","f","g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v","w","x","y","z"]
punctation = [".",",",";",":","?","!","%","$","£","€","#","\\","\"","&","~","{","(","[","`","\\","_","@",")","]","}","=","+","*","\/","<",">",")","}"]
           
liste_fichiers = []
liste_resultats = defaultdict(list)
              
doctype = {"a":"texte",
               "b":"manuscrit",
               "c":"partition",
               "d":"partition manuscrite",
               "e":"carte",
               "f":"carte manuscrite",
               "g":"video",
               "i":"son - non musical",
               "j":"son - musique",
               "k":"image, dessin",
               "l":"ressource électronique",
               "m":"multimedia",
               "r":"objet"
               }
recordtype = {"a":"analytique",
               "i":"feuillets mobiles, etc",
               "m":"monographie",
               "s":"périodiques",
               "c":"collection"}

doc_record_type = defaultdict(str)
for doct in doctype:
    for recordt in recordtype:
        dcrec = doct + recordt
        dcrec_libelles = "-".join([doctype[doct],recordtype[recordt]])
        doc_record_type[dcrec] = dcrec_libelles
#suppression des signes de ponctuation
def clean_punctation(text):
    for char in punctation:
        text = text.replace(char, " ")
    return text

def clean_letters(text):
    for char in letters:
        text = text.replace(char, " ")
    return text

def clean_spaces(text):
    text = re.sub("\s\s+" , " ", text).strip()
    return text

def clean_accents_case(text):
    text = unidecode(text).lower()
    return text
# =============================================================================
# Définition des zones pour chaque élément
# =============================================================================

def record2doctype(label):
    return label[6]

def record2recordtype(label):
    return label[7]

def path2value(record, field_subfield):
    value = None
    val_list = []
    #print(field_subfield)
    if (field_subfield.find("$")>-1):
        field = field_subfield.split("$")[0]
        subfield = field_subfield.split("$")[1]
        for f in record.get_fields(field): 
            for subf in f.get_subfields(subfield):
                val_list.append(subf)
        if (val_list != []):
            value = ";".join(val_list)
    else:
        if (record[field_subfield] is not None and int(field_subfield) < 10):
            value = record[field_subfield].data
    return value

def RepresentsInt(s):
    try: 
        int(s)
        return True
    except ValueError:
        return False


def record2meta(record, liste_elements, alternate_list=[]):
    zone = []
    for el in liste_elements:
        value = path2value(record, el)
        #print("record2meta : " + el + " / "  + str(value))
        if (value is not None):
            zone.append(value)
    #zone = [path2value(record, el) for el in liste_elements if path2value(record, el) is not None]
    if (zone == [] and alternate_list != []):
        for el in alternate_list:
            value = path2value(record, el)
            if (value is not None):
                zone.append(value)
        #zone = [path2value(record, el) for el in alternate_list if path2value(record, el) is not None]
    zone = " ".join(zone)
    #print(zone)
    return zone

def record2title(f200a_e):
    title = clean_spaces(f200a_e)
    title = clean_punctation(title)
    return title

def record2date(f100a,f210d):
    date = ""
    if (RepresentsInt(f100a[9:13]) == True):
        date = f100a[9:13]
    else:
        date = f210d
    date = clean_punctation(date)
    date = clean_letters(date)
    date = clean_spaces(date)
    return date

def record2authors(value_fields):
    authors = clean_spaces(value_fields).strip()
    authors = clean_punctation(authors)
    return authors
    
def aut2keywords(authors):
    authors = clean_punctation(authors)
    liste_authors = authors.split(" ")
    liste_authors = [el for el in liste_authors if el != ""]
    authors2keywords = set()
    for mot in liste_authors:
        authors2keywords.add(clean_accents_case(mot).strip())
    authors2keywords = " ".join(list(authors2keywords))
    return authors2keywords

def record2ark(f033a):
    ark = ""
    if (f033a.find("ark:/12148/") > -1):
        ark = f033a
    return ark

def record2frbnf(f035a):
    frbnf = []
    f035a = f035a.lower().split(";")
    for f035 in f035a:
        if (f035.find("frbn")>-1):
            frbnf.append(f035)
    frbnf = ";".join(frbnf)
    return frbnf

def record2isbn(f010a):
    isbn = f010a
    return isbn

def record2ean(f038a):
    ean = f038a
    return ean

def record2pubPlace(f210a):
    pubPlace = f210a
    return pubPlace

def record2publisher(f210c):
    publisher = clean_spaces(f210c)
    publisher = clean_punctation(publisher)
    return publisher

def record2numeroTome(f200h):
    numeroTome = f200h
    return numeroTome

def record2issn(f011a):
    issn = f011a
    return issn

def record2id_commercial_aud(f073a):
    return f073a
# =============================================================================
# Gestion des mises à jour
# =============================================================================
def download_last_update():
    url = "https://github.com/Transition-bibliographique/alignements-donnees-bnf/"
    webbrowser.open(url)


def iso2tables(master,entry_filename, id_traitement):
    with open(entry_filename, 'rb') as fh:
        collection = mc.MARCReader(fh)
        collection.force_utf8 = True
        try:
            for record in collection:
                record2listemetas(record)
        except mc.exceptions.RecordLengthInvalid:
            print("\n\n/*---------------------------------------------*\n\n")
            print(main.errors["pb_input_utf8"])
            print("\n\n*------------------------------------------------*/")
            main.popup_errors(master,main.errors["pb_input_utf8"])
        

def xml2tables(master,entry_filename, id_traitement):
    collection = mc.marcxml.parse_xml_to_array(entry_filename, strict=False)
    for record in collection:
        record2listemetas(record)


def record2listemetas(record):
    numNot = record2meta(record,["001"])
    doctype = record2doctype(record.leader)
    recordtype = record2recordtype(record.leader)
   
    title = record2title(
                record2meta(record, ["200$a","200$e"])
                )
    keyTitle = record2title(
                record2meta(record, ["530$a"], ["200$a","200$e"])
                )
    authors = record2authors(record2meta(record, [
                                    "700$a",
                                    "700$b",
                                    "710$a",
                                    "710$b",
                                    "701$a",
                                    "701$b",
                                    "711$a",
                                    "711$b",
                                    "702$a",
                                    "702$b",
                                    "712$a",
                                    "712$b"
                                    ],
                        ["200$f"])
                        )
    authors2keywords = aut2keywords(authors)
    date = record2date(record2meta(record,["100$a"]), record2meta(record,["210$d"], ["219$d", "219$i", "219$p"]))
    numeroTome = record2numeroTome(record2meta(record,["200$h"], ["461$v"]))
    publisher = record2publisher(record2meta(record,["210$c"]))
    pubPlace = record2pubPlace(record2meta(record,["210$a"]))
    ark = record2ark(record2meta(record,["033$a"]))
    frbnf = record2frbnf(record2meta(record,["035$a"]))
    isbn = record2isbn(record2meta(record,["010$a"]))
    issn = record2isbn(record2meta(record,["011$a"]))
    ean =  record2ean(record2meta(record,["038$a"]))
    id_commercial_aud = record2id_commercial_aud(record2meta(record,["073$a"]))
    

    doc_record = doctype + recordtype
    doc_record = doc_record.strip()
    if (doc_record not in liste_fichiers):
        liste_fichiers.append(doc_record)
        
    meta = []
    if (doc_record == "am"):
        meta = [numNot,frbnf,ark,isbn,ean,title,authors, date, numeroTome, publisher]
    elif (doc_record == "jm"):
        meta = [numNot,frbnf,ark,ean,id_commercial_aud,title,authors2keywords,date, publisher]
    elif (doc_record == "as"):
        if (keyTitle == ""):
            meta = [numNot, frbnf, ark, issn, title, authors, date, pubPlace]
        else:
            meta = [numNot, frbnf, ark, issn, keyTitle, authors, date, pubPlace]
    else:
        meta = [numNot, frbnf, ark, ean, title, authors, date]
    liste_resultats[doc_record].append(meta)
            
def write_reports(id_traitement):
    for doc_record in liste_resultats:
        file = create_file_doc_record(doc_record, id_traitement)
        file.write("\t".join(["NumNotice","FRBNF","ARK","Autres métadonnées..."]) + "\n")
        for record in liste_resultats[doc_record]:
            print(doc_record, ' - ', record[0])
            file.write("\t".join(record) + "\n")            

    
def end_of_treatments(form,id_traitement):
    write_reports(id_traitement)
    form.destroy()


def launch(form,entry_filename,file_format, output_ID,master):
    main.check_file_name(form,entry_filename)
    print("Fichier en entrée : ", entry_filename)
    if (file_format == 1):
        iso2tables(master,entry_filename, output_ID)
    else:
        xml2tables(master,entry_filename, output_ID)
    end_of_treatments(form,output_ID)



def formulaire_marc2tables(master,access_to_network=True, last_version=[version, False]):
# =============================================================================
# Structure du formulaire - Cadres
# =============================================================================
    couleur_fond = "white"
    couleur_bouton = "#2D4991"
    #couleur_bouton = "#99182D"
    
    [form,
     zone_alert_explications,
     zone_access2programs,
     zone_actions,
     zone_ok_help_cancel,
     zone_notes] = main.form_generic_frames(master,"Conversion de fichiers de notices MARC en tableaux",
                                      couleur_fond,couleur_bouton,
                                      access_to_network)
    
    
    cadre_input = tk.Frame(zone_actions, highlightthickness=2, highlightbackground=couleur_bouton, relief="groove", height=150, padx=10,bg=couleur_fond)
    cadre_input.pack(side="left", anchor="w")
    cadre_input_header = tk.Frame(cadre_input,bg=couleur_fond)
    cadre_input_header.pack(anchor="w")
    cadre_input_file = tk.Frame(cadre_input,bg=couleur_fond)
    cadre_input_file.pack(anchor="w")
    cadre_input_file_name = tk.Frame(cadre_input_file,bg=couleur_fond)
    cadre_input_file_name.pack(side="left")
    cadre_input_file_browse = tk.Frame(cadre_input_file,bg=couleur_fond)
    cadre_input_file_browse.pack(side="left")
    cadre_input_infos_format = tk.Frame(cadre_input,bg=couleur_fond)
    cadre_input_infos_format.pack(anchor="w")
    cadre_input_type_docs = tk.Frame(cadre_input,bg=couleur_fond)
    cadre_input_type_docs.pack(anchor="w")
    
    cadre_inter = tk.Frame(zone_actions, borderwidth=0, padx=10,bg=couleur_fond)
    cadre_inter.pack(side="left")
    tk.Label(cadre_inter, text=" ",bg=couleur_fond).pack()


#=============================================================================
#     Formulaire - Fichier en entrée
# =============================================================================
 
    cadre_output = tk.Frame(zone_actions, highlightthickness=2, highlightbackground=couleur_bouton, relief="groove", height=150, padx=10,bg=couleur_fond)
    cadre_output.pack(side="left")
    cadre_output_header = tk.Frame(cadre_output,bg=couleur_fond)
    cadre_output_header.pack(anchor="w")
    cadre_output_nom_fichiers = tk.Frame(cadre_output,bg=couleur_fond)
    cadre_output_nom_fichiers.pack(anchor="w")
    cadre_output_explications = tk.Frame(cadre_output, padx=20,bg=couleur_fond)
    cadre_output_explications.pack(anchor="w")
    
    cadre_valider = tk.Frame(zone_ok_help_cancel, borderwidth=0, relief="groove", height=150, padx=10,bg=couleur_fond)
    cadre_valider.pack(side="left")
    
    #définition input URL (u)
    tk.Label(cadre_input_header,bg=couleur_fond, fg=couleur_bouton, text="En entrée :", justify="left", font="bold").pack(anchor="w")
    
    tk.Label(cadre_input_file_name,bg=couleur_fond, text="Fichier contenant les notices : ").pack(side="left")
    entry_filename = tk.Entry(cadre_input_file, width=40, bd=2)
    entry_filename.pack(side="left")
    entry_filename.focus_set()
    
    
    #tk.Button(cadre_input_file_browse, text="Sélectionner le fichier\ncontenant les notices", command=lambda:main.openfile(cadre_input_file_name, popup_filename), width=20).pack()
    
    
    tk.Label(cadre_input_type_docs,bg=couleur_fond, text="\nFormat", 
             anchor="w", justify="left").pack(anchor="w")
    file_format = tk.IntVar()
    tk.Radiobutton(cadre_input_type_docs,bg=couleur_fond, text="iso2709", variable=file_format, value=1,
                   anchor="w",justify="left").pack(anchor="w")
    tk.Radiobutton(cadre_input_type_docs,bg=couleur_fond, text="Marc XML", variable=file_format, value=2,
                   anchor="w", justify="left").pack(anchor="w")
    file_format.set(1)
    
    tk.Label(cadre_input_type_docs, text="\n\n\n", bg=couleur_fond).pack()
    
# =============================================================================
#     Formulaire - Fichiers en sortie
# =============================================================================
# 
    
    #Choix du format
    tk.Label(cadre_output_header,bg=couleur_fond, fg=couleur_bouton, font="bold", 
             text="En sortie :", 
             justify="left").pack()
    tk.Label(cadre_output_nom_fichiers,bg=couleur_fond, 
             text="Identifiant des fichiers en sortie : ", 
             justify="left").pack(side="left")
    output_ID = tk.Entry(cadre_output_nom_fichiers, width=40, bd=2)
    output_ID.pack(side="left")
    
    #Ajout (optionnel) d'un identifiant de traitement
    message_fichiers_en_sortie = """
  Le programme va générer plusieurs fichiers, par type de document,
  en fonction du processus d'alignement avec les données de la BnF
  et des métadonnées utilisées pour cela :
        - monographies imprimées
        - périodiques
        - audiovisuel (CD/DVD)
        - autres non identifiés
  Pour faire cela, il utilise les informations en zones codées dans chaque notice Unimarc
    """
    tk.Label(cadre_output_explications,bg=couleur_fond, 
             text=message_fichiers_en_sortie,
             justify="left").pack()
    #explications.pack()
    
    
    
    
    #Bouton de validation
    
    b = tk.Button(cadre_valider, bg=couleur_bouton, fg="white", font="bold", text = "OK", 
                  command=lambda: launch(form, entry_filename.get(), file_format.get(), output_ID.get(),master), 
                  borderwidth=5 ,padx=10, pady=10, width=10, height=4)
    b.pack()
    
    tk.Label(cadre_valider, font="bold", text="", bg=couleur_fond).pack()
    
    call4help = tk.Button(cadre_valider, text="Besoin d'aide ?", command=lambda: main.click2help("https://github.com/Transition-bibliographique/alignements-donnees-bnf/"), padx=10, pady=1, width=15)
    call4help.pack()
    
    cancel = tk.Button(cadre_valider, bg=couleur_fond, text="Annuler", command=lambda: main.annuler(form), padx=10, pady=1, width=15)
    cancel.pack()
    
    tk.Label(zone_notes, text = "Version " + str(main.version) + " - " + lastupdate, bg=couleur_fond).pack()
    
    """if (main.last_version[1] == True):
        download_update = tk.Button(zone_notes, text = "Télécharger la version " + str(main.last_version[0]), command=download_last_update)
        download_update.pack()"""

    tk.mainloop()

if __name__ == '__main__':
    access_to_network = main.check_access_to_network()
    #if(access_to_network is True):
    #    last_version = main.check_last_compilation(programID)
    main.formulaire_main(access_to_network, last_version)
    #formulaire_marc2tables(access_to_network,last_version)

